website/source/guides/writing-custom-terraform-providers.html.md

```diff
@@ -0,0 +1,553 @@
+contents of the main function consume Terraform's `plugin` library. This library
+deals with all the communication between Terraform core and the plugin.
+
+Next, build the plugin using the Go toolchain:
+
+```shell
+$ go build -o terraform-provider-example
+```
+
+The output name (`-o`) is **very important**. Terraform searches for plugins in
+the format of:
+
+```text
+terraform-<TYPE>-<NAME>
+```
+
+In the case above, the plugin is of type "provider" and of name "example".
+
+To verify things are working correctly, execute the binary just created:
+
+```shell
+$ ./terraform-provider-example
+This binary is a plugin. These are not meant to be executed directly.
+Please execute the program that consumes these plugins, which will
+load any plugins automatically
+```
+
+This is the basic project structure and scaffolding for a Terraform plugin.
+
+## Defining Resources
+
+Terraform providers manage resources. A provider is an abstraction of an
+upstream API, and a resource is a component of that provider. As an example, the
+AWS provider supports `aws_instance` and `aws_elastic_ip`. DNSimple supports
+`dnsimple_record`. Fastly supports `fastly_service`. Let's add a resource to our
+fictitious provider.
+
+As a general convention, Terraform providers put each resource in their own
+file, named after the resource, prefixed with `resource_`. To create an
+`example_server`, this would be `resource_server.go` by convention:
+
+```go
+package main
+
+import (
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceServer() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceServerCreate,
+		Read:   resourceServerRead,
+		Update: resourceServerUpdate,
+		Delete: resourceServerDelete,
+
+		Schema: map[string]*schema.Schema{
+			"address": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+```
+
+This uses the [`schema.Resource`
+type](https://godoc.org/github.com/hashicorp/terraform/helper/schema#Resource).
+This structure defines the data schema and CRUD operations for the resource.
+Defining these properties are the only required thing to create a resource.
+
+The schema above defines one element, `"address"`, which is a required string.
+Terraform's schema automatically enforces validation and type casting.
+
+Next there are four "fields" defined - `Create`, `Read`, `Update`, and `Delete`.
+These four functions are required for a resource to be functional. There are
+other functions, but these are the only required ones. Terraform itself handles
+which function to call and with what data. Based on the schema and current state
+of the resource, Terraform can determine whether it needs to create a new
+resource, update an existing one, or destroy.
+
+Each of the four struct fields point to a function. While it is technically
+possible to inline all functions in the resource schema, best practice dictates
+pulling each function into its own method. This optimizes for both testing and
```

Comment:
Might mention to pull each function into it's own method with the exception of the schema key